### PR TITLE
feat(robot_server): add system/time GET & PUT endpoints

### PR DESCRIPTION
**Pull request by sanni-t**
# Overview

This PR adds GET & PUT time endpoints to robot server. These will be used by the RunApp to get/ update robot's date & time. This will enable us to update robot's time in absence of an RTC/ internet access. 
Addresses #3872 

# Changelog

- added `system` service to `robot_server`
- added routers, models & errors for the time endpoints in `/system`
- added `time` component to `opentrons/system`
- added api & robot_server tests

# Review requests

- Check for code sensibility
- Do the tests make sense? Any tests missing?
- Upload to a robot & test the endpoints!
   - If your robot is ntp synchronized, you should NOT be able to update system time using PUT
   - If your robot has not had ntp synchronization since boot, PUT /system/time should successfully set system time
   - Do the responses make sense?
   - All responses should be in UTC (+00:00 offset)
   - Input time value, regardless of its timezone specification,  should set the correct UTC time on robot

# Risk assessment

Low. The endpoints are not being used by anything right now & the code is isolated.

### CODE REVIEW

File: api/src/opentrons/system/time.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        f'date --utc --set \"{time}\"',
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+        loop=loop or asyncio.get_event_loop()
+    )
+    out, err = await proc.communicate()
+    return out.decode(), err.decode()
+
+
+async def get_system_time(loop: asyncio.AbstractEventLoop = None) -> datetime:
```
Comment (sfoster1) [inline api/src/opentrons/system/time.py:49]: This function is going to be called pretty frequently so the app can make sure that the time stays in sync, right? I don't think we should be shelling out for something like that, or that frequently. `datetime.datetime.now()` should work a lot better with just a single syscall and file access.

File: api/src/opentrons/system/time.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    else:
+        new_time_dt = new_time_dt.astimezone(tz=timezone.utc)
+        new_time_str = new_time_dt.strftime("%Y-%m-%d %H:%M:%S")
+        log.info(f'Setting time to {new_time_str} UTC')
+        out, err = await _set_time(new_time_str)
+        log.info(f'{out if out else err}')
+    now = await get_system_time(loop)
+    # TODO: check with team whether to return whole error or
+    #  just an error status
+    return now, err
```
Comment (sfoster1) [inline api/src/opentrons/system/time.py:80]: we can definitely log the error message if it exists at `error` level (unless that gets too spammy)

File: api/src/opentrons/system/time.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    status = await _time_status(loop)
+    if status['LocalRTC'] == "yes" or status['NTPSynchronized'] == "yes":
+        # TODO: Update this to handle RTC sync correctly once we introduce RTC
+        err = 'Cannot set system time; already synchronized with NTP or RTC'
+    else:
+        new_time_dt = new_time_dt.astimezone(tz=timezone.utc)
+        new_time_str = new_time_dt.strftime("%Y-%m-%d %H:%M:%S")
+        log.info(f'Setting time to {new_time_str} UTC')
+        out, err = await _set_time(new_time_str)
+        log.info(f'{out if out else err}')
```
Comment (sfoster1) [inline api/src/opentrons/system/time.py:76]: This might be a little spammy

File: robot-server/robot_server/service/system/models.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from datetime import datetime
+from pydantic import BaseModel, Field
+
+from robot_server.service.json_api import ResponseModel, ResponseDataModel, \
+    RequestDataModel, RequestModel
+
+
+class SystemTimeAttributes(BaseModel):
+    systemTime: datetime
+    setTimeError: Optional[str]
```
Comment (sfoster1) [inline robot-server/robot_server/service/system/models.py:11]: Should we do this or should we use an error response?

File: api/src/opentrons/system/time.py (hunk 1 above)
Comment (sanni-t) [reply api/src/opentrons/system/time.py:49]: No really. That was indeed what we had drafted, but after giving it some thought, we decided to update system time only when user initiates connection to the robot. 
Although, that does not take into account any time getters in order to show robot time in the app, if we decide to do that. In which case, you are right, we'll need to switch to `datetime.now()`.

Comment (sanni-t) [reply api/src/opentrons/system/time.py:49]: Thought about it a bit more, you're right, even if we won't be calling the function frequently right now, there's no harm in making the switch to `datetime.now()`. That way we won't have make the change later.

File: robot-server/robot_server/service/system/models.py (hunk 4 above)
Comment (sanni-t) [reply robot-server/robot_server/service/system/models.py]: Good point. Changed to error response.

**Comment by sanni-t** [comment]
> It kinda seems like `opentrons.system.time` should be `robot_server.(something).time`. None of its functions are called from the rest of the `opentrons` module, and it isn't really related to robot operation. I know there's other system components in the `opentrons` package, but that's just because the `robot_server` package didn't exist when they were written, and I think we should move them as well (not necessarily in this pr though)

Agreed about moving system components out of opentrons module and into robot_server. Since there's other components that need to be moved too, I'll put up a different PR to move all of them at once.

File: api/src/opentrons/system/time.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+from datetime import datetime, timezone
+from dateutil.parser import parse
+
+
+log = logging.getLogger(__name__)
+
+
+def _str_to_dict(res_str):
+    lines = res_str.split('\n')
+    res_dict = {}
```
[the remark this answers is not in the document]
Comment (sanni-t) [reply api/src/opentrons/system/time.py:12]: Converting the string items to Dataclass objects seemed a bit too cumbersome esp when the objects would be used just once in the code. 
Converted the the dict yes/no items to bools

File: api/src/opentrons/system/time.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+def _str_to_dict(res_str) -> Dict[str, Union[str, bool]]:
+    res_lines = res_str.splitlines()
+    res_dict = {}
+    try:
+        for line in res_lines:
+            if line:
+                prop, val = line.split('=')
+                res_dict[prop] = val if val not in ['yes', 'no'] \
+                    else val == 'yes'  # Convert yes/no to boolean value
+    except (ValueError, IndexError) as e:
```
Comment (sfoster1) [inline api/src/opentrons/system/time.py:19]: [nit]: the most probably failure case here is that we didn't anticipate a particular kind of output from `timedatectl`, but the output we _did_ anticipate is the same. given that, should we maybe do this kind of checking on a line-by-line basis and not raise the exception? instead, we can log with a specific tag and look for it on log monitoring.

File: robot-server/robot_server/service/system/router.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+@router.put("/system/time",
+            description="Update system time",
+            summary="Set robot time",
+            response_model=time_models.SystemTimeResponse)
+async def set_time(new_time: time_models.SystemTimeRequest) \
+        -> time_models.SystemTimeResponse:
+    sys_time, err = await time.set_system_time(
+        new_time.data.attributes.systemTime)
+    if err:
+        if 'already synchronized with NTP or RTC' in err:
```
Comment (sfoster1) [inline robot-server/robot_server/service/system/router.py:63]: To me, this is the wrong place to do this. It leaks implementation details - the robot server endpoint function now has to care about the exact details of how `timesyncd` formats error messages.

File: robot-server/tests/service/system/test_router.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    })
+    assert response.json() == {'errors': [{
+        'detail': 'Something went wrong.',
+        'status': '500',
+        'title': 'Internal Server Error'}]}
+    assert response.status_code == 500
+
+
+def test_get_system_time(api_client, mock_system_time, response_links):
```
Comment (sfoster1) [inline robot-server/tests/service/system/test_router.py:68]: This probably wants to be a tavern integration test

File: robot-server/tests/service/system/test_router.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            'attributes': {'systemTime': mock_system_time.isoformat()},
+            'id': 'time',
+            'type': 'SystemTimeAttributes'},
+        'links': response_links,
+        'meta': None
+    }
+    assert response.status_code == 200
+
+
+def test_set_with_missing_field(api_client, mock_system_time):
```
Comment (sfoster1) [inline robot-server/tests/service/system/test_router.py:85]: This probably wants to be a tavern integration test

File: robot-server/tests/service/system/test_router.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    assert response.json() == {
+        'errors': [{
+            'detail': 'field required',
+            'source': {'pointer': '/body/new_time'},
+            'status': '422',
+            'title': 'value_error.missing'}]}
+    assert response.status_code == 422
+
+
+def test_set_system_time(api_client, mock_system_time, response_links):
```
Comment (sfoster1) [inline robot-server/tests/service/system/test_router.py:97]: this _probably_ wants to be a tavern integration test, but given the degree of mocking maybe it wouldn't be worth it.

**Comment by sanni-t** [comment]
@sfoster1 Made all the changes requested. Talked with Amit offline about tavern tests and we decided to only add those tests that don't require mocking the system components.

File: robot-server/robot_server/service/system/router.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+            description="Fetch system time & date",
+            summary="Get robot's time status, which includes- current UTC "
+                    "date & time, local timezone, whether robot time is synced"
+                    " with an NTP server &/or it has an active RTC.",
+            response_model=time_models.SystemTimeResponse
+            )
+async def get_time() -> time_models.SystemTimeResponse:
+    res = await time.get_system_time()
+    return _create_response(res)
+
```
Comment (amitlissack) [inline robot-server/robot_server/service/system/router.py:51]: This warning is not quite accurate. We need to configure tests to recognize coverage of tavern tests. I will make a ticket now.

File: robot-server/robot_server/system/time.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    res_dict = {}
+
+    for line in res_lines:
+        if line:
+            try:
+                prop, val = line.split('=')
+                res_dict[prop] = val if val not in ['yes', 'no'] \
+                    else val == 'yes'  # Convert yes/no to boolean value
+            except (ValueError, IndexError) as e:
+                log.error("Error converting timedatectl string: {}".format(e))
```
Comment (amitlissack) [inline robot-server/robot_server/system/time.py:22]: I think it's worth logging the actual line contents too.

**Review by amitlissack — APPROVED** [review]
Looks good to me!

**Review by mcous — APPROVED** [review]

File: robot-server/robot_server/service/system/router.py (hunk 11 above)
Comment (sfoster1) [reply robot-server/robot_server/service/system/router.py:51]: That's a really cool feature to display it inline though!

**Review by sfoster1 — APPROVED** [review]
Looks good to me! Love it being integrated into the robot server

**Pull request merged**